docs/attributes-registry/db.md

```diff
@@ -21,16 +19,17 @@
 | Attribute  | Type | Description  | Examples  | Stability |
 |---|---|---|---|---|
 | `db.collection.name` | string | The name of a collection (table, container) within the database. [1] | `public.users`; `customers` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| `db.collection.namespace` | string | The namespace containing database objects, fully qualified within the server address and port. [2] | `customers`; `test.users` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | `db.instance.id` | string | An identifier (address, unique name, or any other identifier) of the database instance that is executing queries or mutations on the current connection. This is useful in cases where the database is running in a clustered environment and the instrumentation is able to record the node executing the query. The client may obtain this value in databases like MySQL using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| `db.name` | string | This attribute is used to report the name of the database being accessed. For commands that switch the database, this should be set to the target database (even if the command fails). [2] | `customers`; `main` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

Comment:
Is there a way we can add `db.collection.namespace` without getting rid of `db.name` entirely? I recognize there are gaps with `db.name` for some systems, but this is an intuitive name that covers the simple use case well.

Reply:
sure, I think it's worth reconsidering the name `db.collection.namespace` given we are no longer thinking to parse the namespace from `db.collection.name`

in Java, for example, `db.collection.namespace` will probably be captured via some combination of

  * [java.sql.Connection#getCatalog()](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getCatalog%E2%80%93)
  * [java.sql.DatabaseMetaData.getUserName()](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getUserName--)
    * in some databases (e.g. PostgreSQL and Oracle), users are owners of database objects and so they are part of an object's namespace
  * [DatabaseMetaData.getURL()](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getURL--)
    * this could be used to parse out some parts of the "namespace", e.g. the Oracle Instance Name

and these aren't really scoped to the collection in `db.collection.name`, so maybe we shouldn't scope namespace to `db.collection.*`

Reply:
@jcocchi do I understand correctly, for CosmosDB there are 3 "layers"
* Account - this would be captured in `server.address`?
* Database - this would be captured in the "namespace" attribute we're discussing?
* Collection - this would be captured in `db.collection.name`?

Reply:
> * [java.sql.Connection#getCatalog()](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getCatalog%E2%80%93)
> * [java.sql.DatabaseMetaData.getUserName()](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getUserName--)

btw, I don't know
 if these make network calls, and so whether they would be practical to use from instrumentation

Reply:
The catalog seems to be static and local:
- postgres [1](https://github.com/pgjdbc/pgjdbc/blob/d5cfb04d1413685ce4d4a9a2f641d88b514bcdaf/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1115) and [2](https://github.com/pgjdbc/pgjdbc/blob/d5cfb04d1413685ce4d4a9a2f641d88b514bcdaf/pgjdbc/src/main/java/org/postgresql/core/QueryExecutorBase.java#L141))
- [SQL Server](https://github.com/microsoft/mssql-jdbc/blob/193edbdaf085786fbe47227e80b88d942e0547ad/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java#L4900)
- [MySQL](https://github.com/mysql/mysql-connector-j/blob/805f872a57875f311cb82487efcfb070411a3fa0/src/main/user-impl/java/com/mysql/cj/jdbc/ConnectionImpl.java#L1088)
- Oracle: gave up reading decompiled jar

User - mostly dynamic and requires network call:
- [SQL Server](https://github.com/microsoft/mssql-jdbc/blob/main/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java#L1784) - dynamic
- [Postgres](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1671) - dynamic
- [MySQL](https://github.com/mysql/mysql-connector-j/blob/805f872a57875f311cb82487efcfb070411a3fa0/src/main/user-impl/java/com/mysql/cj/jdbc/DatabaseMetaData.java#L4175) - sometimes static sometimes dynamic depending on the config. Seems to be dynamic by default

Schema - mostly dynamic and requires network call:
- [Postgres](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1672) - dynamic
- [SQL Server](https://github.com/microsoft/mssql-jdbc/blob/193edbdaf085786fbe47227e80b88d942e0547ad/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java#L7271) - dynamic
- [MySQL](https://github.com/mysql/mysql-connector-j/blob/805f872a57875f311cb82487efcfb070411a3fa0/src/main/user-impl/java/com/mysql/cj/jdbc/ConnectionImpl.java#L2393) static/local and the same as database name (depends on mode)

Reply:
For Oracle JDBC driver:

Catalog - [always null](https://github.com/caot/ojdbc6-11.2.0.2.0.src/blob/41eef5e34775532c7aec2f03d41d360152676e00/src/oracle/jdbc/driver/PhysicalConnection.java#L3230-L3234)

User - [cached after the first call](https://github.com/caot/ojdbc6-11.2.0.2.0.src/blob/41eef5e34775532c7aec2f03d41d360152676e00/src/oracle/jdbc/driver/PhysicalConnection.java#L4042-L4049)

Reply:
I think this part of https://stackoverflow.com/questions/7942520/relationship-between-catalog-schema-user-and-database-instance is interesting:

to identify any object in running server, you need

* Oracle - schema name + object name
* PostgreSQL - database name + schema name + object name
* MySQL - database name + object name
* SQLServer - database name + schema name + object name

to identify the running server itself

* Oracle - [instance name](https://docs.oracle.com/en/database/oracle/oracle-database/21/jajdb/)
* PostgreSQL - n/a
* MySQL - n/a
* SQLServer - [instance name](https://learn.microsoft.com/en-us/sql/connect/jdbc/building-the-connection-url?view=sql-server-ver16)

so we have these 4 things:

* instance name - doesn't change after connection, can be captured from connection string (probably maps to `db.instance.id`, maybe could consider renaming to `db.instance.name`)
* database name - can change after connection, JDBC drivers seem to track these changes and expose via getCatalog, but could be harder to capture in some languages
* schema name - can change after connection, JDBC drivers don't appear to track these changes, so probably cannot capture in the generic case
* object name - `db.collection.name`

Reply:
it doesn't seem that the above JDBC drivers track "use <database>" calls, and so getCatalog() could be incorrect, though I guess it could be said that best practice is to call `setCatalog(<database>)`, which the drivers do track

Reply:
I updated the note on the SQL namespace to be

```yaml
- ref: db.collection.namespace
  brief: The namespace containing database objects, fully qualified within the server address and port.
  note: >
    Namespace contains different identifiers depending on the database system:

    * MySQL - database or schema name. For example, `products.dbo`
    * PostgreSQL - database and schema name. For example, `products.public`
    * MS SQL Server - instance name, database and schema name. For example, `test.products.public`
    * Oracle - instance name and schema name. For example, `test.products`

    Current database name can usually be obtained using database driver API such as
    [JDBC `Connection.getCatalog()`](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getCatalog--)
    or [.NET `SqlConnection.Database`](https://learn.microsoft.com/dotnet/api/system.data.sqlclient.sqlconnection.database).

    If instrumentation cannot reliably determine the database name (for example, if database can be changed
    in runtime without instrumentation being aware of it), it SHOULD NOT set `db.collection.namespace`.

    Instrumentations that parse SQL statements MAY use database name provided in the connection string
    and track the currently selected database name as long as current database name is associated with
    current connection and can't be changed without instrumentation being aware of it.
 
    For commands that switch the database, this should be set to the target database (even if the command fails).

    TODO schema.
```

From here, I see the following options:
1. allow to skip schema if it's not known.
    - i.e. MS SQL Server would have `db.collection.namespace = instance.db` ~~most~~all of the time. 
    - Oracle might need the whole attribute to be opt-in to make a network call to get the current schema and then monitor `ALTER SESSION SET CURRENT_SCHEMA` calls
2. don't include schema into the namespace at all. 
    - it could be bad for Oracle, but all options are bad for it.
3. Gave up on the idea to have namespace attribute and just do cosmetic improvements to the DB name. I still see the value in the namespace hirarchy though:
    - place for the instance-within-host identifier
    - possible place for [cluster](https://github.com/open-telemetry/semantic-conventions/issues/858)/shard/partition-id 

Let's talk more about it on the WG call.

Reply:
Added examples for Cosmos and Mongo inline

          * Cosmos DB - database name. For example, `products`
          * MongoDB - database name. For example, `products`

Reply:
any objection to marking this discussion as resolved?

Reply:
Based on the previous discussion (Wed 4/17 DB WG meeting) with @jcocchi I believe `db.namespace` was a good solution and we can resolve this discussion.